File: expo-audio-streaming/ios/AudioStreamingModule.swift
import AVFoundation
import ExpoModulesCore

public class AudioStreamingModule: Module {
  private var audioEngine: AVAudioEngine?
  private var inputNode: AVAudioInputNode?
  private var isRecording = false
  private let sampleRate: Double = 16000

  public func definition() -> ModuleDefinition {
    Name("AudioStreamingModule")

    AsyncFunction("startRecording") { (promise: Promise) in
      do {
        if self.isRecording {
          promise.resolve(nil)
          return
        }

        // Request microphone permission
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setCategory(.record, mode: .default)
        try audioSession.setActive(true)

        // Check permission
        let permissionStatus = audioSession.recordPermission
        if permissionStatus == .denied {
          promise.reject("PERMISSION_DENIED", "Microphone permission not granted", nil)
          return
        }

        if permissionStatus == .undetermined {
          audioSession.requestRecordPermission { granted in
            if granted {
              DispatchQueue.main.async {
                self.startAudioEngine(promise: promise)
              }
            } else {
              promise.reject("PERMISSION_DENIED", "Microphone permission denied", nil)
            }
          }
          return
        }

        // Permission already granted
        startAudioEngine(promise: promise)
      } catch {
        promise.reject("START_FAILED", "Failed to start recording: \(error.localizedDescription)", error)
      }
    }

    AsyncFunction("stopRecording") { (promise: Promise) in
      do {
        self.isRecording = false
        
        if let audioEngine = self.audioEngine {
          audioEngine.stop()
          self.inputNode?.removeTap(onBus: 0)
        }
        
        self.audioEngine = nil
        self.inputNode = nil
        
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setActive(false)
        
        promise.resolve(nil)
      } catch {
        promise.reject("STOP_FAILED", "Failed to stop recording: \(error.localizedDescription)", error)
      }
    }

    AsyncFunction("isRecording") { (promise: Promise) in
      promise.resolve(self.isRecording)
    }
  }

  private func startAudioEngine(promise: Promise) {
    do {
      let audioEngine = AVAudioEngine()
      let inputNode = audioEngine.inputNode
      
      // Configure audio format
      let format = AVAudioFormat(
        commonFormat: .pcmFormatInt16,
        sampleRate: sampleRate,
        channels: 1,
        interleaved: false
      )
      
      guard let audioFormat = format else {
        promise.reject("INIT_FAILED", "Failed to create audio format", nil)
        return
      }

      // Install tap on input node
      let bufferSize: AVAudioFrameCount = 4096
      inputNode.installTap(onBus: 0, bufferSize: bufferSize, format: audioFormat) { [weak self] buffer, _ in
        guard let self = self, self.isRecording else { return }
        
        // Convert buffer to PCM data
        guard let channelData = buffer.int16ChannelData else { return }
        let channelDataValue = channelData.pointee
        let frameLength = Int(buffer.frameLength)
        
        // Convert Int16 array to Data (little-endian)
        var audioData = Data(capacity: frameLength * 2)
        for i in 0..<frameLength {
          let sample = channelDataValue[i]
          audioData.append(contentsOf: withUnsafeBytes(of: sample.littleEndian) { Array($0) })
        }
        
        // Convert to base64
        let base64Data = audioData.base64EncodedString()
        
        // Send event to JavaScript
        self.sendEvent("onAudioChunk", [
          "data": base64Data,
          "mimeType": "audio/pcm;rate=16000"
        ])
      }

      // Start audio engine
      try audioEngine.start()
      
      self.audioEngine = audioEngine
      self.inputNode = inputNode
      self.isRecording = true
      
      promise.resolve(nil)
    } catch {
      promise.reject("START_FAILED", "Failed to start audio engine: \(error.localizedDescription)", error)
    }
  }
}

File: expo-audio-streaming/android/src/main/java/expo/audiostreaming/AudioStreamingModule.kt
package expo.audiostreaming

import android.Manifest
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.util.Base64
import expo.modules.core.interfaces.ExpoProp
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.isActive
import java.nio.ByteBuffer
import java.nio.ByteOrder

class AudioStreamingModule : Module() {
  private var audioRecord: AudioRecord? = null
  private var isRecording = false
  private var recordingJob: Job? = null
  private val recordingScope = CoroutineScope(Dispatchers.IO)
  
  private val sampleRate = 16000
  private val channelConfig = AudioFormat.CHANNEL_IN_MONO
  private val audioFormat = AudioFormat.ENCODING_PCM_16BIT
  private val bufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat) * 2

  override fun definition() = ModuleDefinition {
    Name("AudioStreamingModule")

    AsyncFunction("startRecording") { promise: Promise ->
      try {
        if (isRecording) {
          promise.resolve(null)
          return@AsyncFunction
        }

        // Check permission
        val permissionManager = appContext.permissions
        val hasPermission = permissionManager?.hasGrantedPermissions(Manifest.permission.RECORD_AUDIO) ?: false

        if (!hasPermission) {
          promise.reject("PERMISSION_DENIED", "Microphone permission not granted", null)
          return@AsyncFunction
        }

        // Create AudioRecord
        audioRecord = AudioRecord(
          MediaRecorder.AudioSource.MIC,
          sampleRate,
          channelConfig,
          audioFormat,
          bufferSize
        )

        if (audioRecord?.state != AudioRecord.STATE_INITIALIZED) {
          promise.reject("INIT_FAILED", "Failed to initialize AudioRecord", null)
          return@AsyncFunction
        }

        audioRecord?.startRecording()
        isRecording = true

        // Start streaming audio chunks
        recordingJob = recordingScope.launch {
          streamAudioChunks()
        }

        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("START_FAILED", "Failed to start recording: ${e.message}", e)
      }
    }

    AsyncFunction("stopRecording") { promise: Promise ->
      try {
        isRecording = false
        recordingJob?.cancel()
        recordingJob = null

        audioRecord?.stop()
        audioRecord?.release()
        audioRecord = null

        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("STOP_FAILED", "Failed to stop recording: ${e.message}", e)
      }
    }

    AsyncFunction("isRecording") { promise: Promise ->
      promise.resolve(isRecording)
    }
  }

  private suspend fun streamAudioChunks() {
    val buffer = ShortArray(bufferSize / 2) // 16-bit = 2 bytes per sample
    
    while (isActive && isRecording) {
      val audioRecord = this@AudioStreamingModule.audioRecord
      if (audioRecord == null || audioRecord.state != AudioRecord.STATE_INITIALIZED) {
        break
      }

      val samplesRead = audioRecord.read(buffer, 0, buffer.size)

      if (samplesRead > 0) {
        // Convert ShortArray to ByteArray
        val byteBuffer = ByteBuffer.allocate(samplesRead * 2)
        byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
        for (sample in buffer.take(samplesRead)) {
          byteBuffer.putShort(sample)
        }
        val audioData = byteBuffer.array()

        // Convert to base64
        val base64Data = Base64.encodeToString(audioData, Base64.NO_WRAP)

        // Send event to JavaScript
        sendEvent("onAudioChunk", mapOf(
          "data" to base64Data,
          "mimeType" to "audio/pcm;rate=16000"
        ))
      }

      // Small delay to prevent overwhelming the event system
      kotlinx.coroutines.delay(50)
    }
  }
}

